docs_: code owners policy (#6506)
status-im/status-go: _docs/policies/codeowners.md

# Code Owners Policy

### Purpose

This policy outlines how code ownership is assigned and ensures that core contributors (CCs) are accountable for 
reviewing and maintaining code in the `status-go` repository.

### Code Owners File

- Code owners MUST be stored in the `.github/CODEOWNERS` file, as per GitHub's [CODEOWNERS documentation](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners).

### Assigning Code Ownership

1. Eligibility
 
- Only CCs MAY be code owners. External contributors are not eligible to be code owners.

2. Nomination
 
- For **existing** code:
  - Any CC MAY nominate themselves as a code owner for any part of the existing code.
  - Nominations SHOULD be discussed and agreed upon with other CCs if there are overlaps or concerns.
- For **new** code:

>   - When introducing a new file/module, CCs SHOULD assign themselves or another CC as a code owner for this file/module. 

Reviewer: But do we need an owner for every new **file** created from now on? Seems excessive to me to manage this at the file level. On the other hand, if a new **module** is being introduced, then I agree with the nomination rule above.

Author: I agree that this might become an overkill, but for some huge modules (e.g. `protocol`) I believe there should be such an option (at least for now). Unfortunately, we can easily add a new part of `Messenger` in a separate file, and it would make sense to add a code owner, while it wouldn't make sense for the whole `protocol` module.

I have no strong opinion on this

Reviewer 2: What about something like

```
When introducing a new module or substantial new component, CCs SHOULD assign ownership. File-level ownership MAY be used where appropriate but is not required.
```

I think per file level changes may be overly cumbersome.

Reviewer: @Samyoul's suggestion above strike a good balance to me.

Author: @ilmotta @Samyoul thank you guys, I've added Sam's suggested text in f1807194a04e5c7641d6bf29c1e7d97e91e9828a.

### Pull Request Reviews

> - At least one of the code owners for the affected files MUST review a PR for it to be eligible for merging.

Reviewer 2: What happens if a PR touches files owned by someone that is unavailable for whatever reason(example, away for a week, left the team , left the org), what's the fallback?

What do you think about something like this:

“If a code owner is unresponsive within a reasonable timeframe, another CC from the relevant area MAY approve the PR in their place, where possible after making a comment in the PR.”

Author: @Samyoul makes sense, thank you 👍  Added in 8e71820c98d27ecd5eb0770150746704723934e6.
I've copied your proposal, but removed the requirement for a comment (let's rely on us developers being reasonable in such actions 😄)

# Code Owners Policy

### Purpose

This policy outlines how code ownership is assigned and ensures that core contributors (CCs) are accountable for 
reviewing and maintaining code in the `status-go` repository.

### Code Owners File

- Code owners MUST be stored in the `.github/CODEOWNERS` file, as per GitHub's [CODEOWNERS documentation](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners).

### Assigning Code Ownership

1. Eligibility
 
- Only CCs MAY be code owners. External contributors are not eligible to be code owners.

2. Nomination
 
- For **existing** code:
  - Any CC MAY nominate themselves as a code owner for any part of the existing code.

>   - Nominations SHOULD be discussed and agreed upon with other CCs if there are overlaps or concerns.

Reviewer 3: > **IF** there are overlaps or concerns.

Can we make 'if' capital? (not sure if the notation allows capitalization for anything except MUST MAY SHOULD...)

Author: Resolved now, `if` is gone.

>   - Nominations SHOULD be discussed and agreed upon with other CCs if there are overlaps or concerns.

Reviewer 2: ```suggestion
  - Nominations SHOULD be discussed and agreed upon with other CCs to prevent and address overlaps or concerns.
```

Author: Nice, thank you 👍

- For **new** code:
  - When introducing a new module or substantial new component, CCs SHOULD assign ownership.
  - File-level ownership MAY be used where appropriate but is not required.

### Pull Request Reviews

- At least one of the code owners for the affected files MUST review a PR for it to be eligible for merging.

> - If a code owner is unresponsive within a reasonable timeframe, another CC from the relevant area MAY approve 

Reviewer 4: - I agree with this section
- I feel this line contradicts the previous one (not sure if we need them to be valid when read individually)
- I couldn't quickly come up with a better proposal

> - If a code owner is unresponsive within a reasonable timeframe, another CC from the relevant area MAY approve 

Reviewer 2: `MAY` in this context reads as "is allowed to", the standard use of `MAY` in policies, RFCs, etc reads as "has the option to".

In this case the use of `MAY` is ambiguous.

We could use "may" to communicate that another cc "is allowed to" approve in the place of the unresponsive code owner.

```suggestion
- If a code owner is unresponsive within a reasonable timeframe, another CC from the relevant area may approve 
```

Or, because a relevant CC is required for any code review we specify that "another CC with context MUST approve".

```suggestion
- If a code owner is unresponsive within a reasonable timeframe, another CC from the relevant area MUST approve 
```

Author: Indeed. I like the lower-case `may` here, not explicitly forcing somebody to review. I doubt someone will abuse this rule.
